draft/authentication-oauth2.md

```diff
@@ -0,0 +1,256 @@
+endpoint as described in the OAuth 2 specification.
+
+- Response Type
+- Client ID
+- Client Secret
+- State
+- Scope
+- Redirect Callback
+- Token
+- Token Type
+
+#### Token endpoint
+
+The token endpoint is used by the client to obtain an access token as
+described by [RFC 6749 section 3.2](http://tools.ietf.org/html/rfc6749#section-3.2).
+This endpoint is identified by the `token` relation identifier.
+
+API Blueprint tooling can use the types of parameters or attributes to
+understand the content being placed within them. Although a user may add
+any additonal types of fields as required by the API:
+
+- Grant
+- Code
+- Redirect Callback
+- Client ID
+- Client Secret
+- Token
+- Token Type
+- Refresh Token
+
+For example, to describe a token endpoint that uses the `POST` HTTP
+method against the `/token` URI you can use the following:
+
+```apib
+### Token [POST /token]
+
++ Relation: token
++ Request
+    + Parameters
+        + `grant_type`: `authorization_code` (Grant)
+        + code (Code)
+        + `redirect_uri` (Redirect Callback)
+        + `client_id` (Client ID)
+        + `client_secret` (Client Secret)
+
++ Response 200 (application/json)
+    + Attributes
+        + `access_token` (Token)
+```
+
+When a parameter is not described in the URI template for the URI, it is
+assumed to be a query parameter.
+
+##### Defining custom request and responses that can return an access token.
+
+You can define any custom flow providing it returns a response including
+a access token attribute.
+
+```apib
+### Token [POST /token]
+
++ Relation: token
++ Request (application/json)
+    + `grant_type`: password (Grant)
+    + username
+    + password
+
++ Response 200 (application/json)
+    + Attributes
+        + `access_token` (Token)
+```
+
+```apib
+### Token [POST /token]
+
++ Relation: token
+
++ Request (application/x-www-form-urlencoded)
+    + Headers:
+        + Authorization: Basic czZCaGRSa3F0MzpnWDFmQmF0M2JW
+
+    + Attributes
+        + `grant_type`: `refresh_token`
+        + `refresh_token` (Refresh Token)
+
++ Response 200 (application/json)
+    + Attributes
+        + `access_token` (Token)
+```
+
+#### Revocation Endpoint
+
+You may define an endpoint that can be used to revoke an access token using the
+`revoke` relation identifier.
+
+```apib
+### Revoke an access token [DELETE /revoke]
+
++ Relation: revoke
++ Response 204
+```
+### Access Token Types
+
+API Blueprint tooling may only support the Bearer token type as per described
+in this RFC, however any tooling may support other token types defined in
+the OAuth specification.
+
+#### Bearer Token
+
+As defined in [RFC6750](http://tools.ietf.org/html/rfc6750), the Bearer token
+may be used in many ways.
+
+Inside an OAuth 2 authentication scheme, you can show requests using a Bearer
+token under the `Bearer` heading.
+
+```apib
+### Bearer
+
++ Request
+    + Headers
+        + Authorization (Token)
```

Comment:
This unfortunately isn't correct but I'm unsure how to clearly express this. I'm wondering if anyone else has any elegant solutions.

Basically, the Bearer token will be in the `Authorization` header. It won't be in the header alone which is why this won't work. It will have a suffix like `Authorization: Bearer f136803ab9c241079ba0cc1b5d02ee77`.